src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,264 @@
+    {
+        private readonly CompactTree _tree;
+        private readonly int _samples;
+        private readonly int _seed;
+        private RandomIterator _iterator;
+
+        public RandomDictionaryKeyScanner(CompactTree tree, int samples, int seed = 0)
+        {
+            _tree = tree;
+            _seed = seed;
+            _samples = samples;
+            _iterator = _tree.RandomIterate(samples, seed: seed);
+            _iterator.Reset();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.RandomIterate(_samples, seed: _seed);
+            _iterator.Reset();
+        }
+    }
+
+    public struct RandomBranchDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly long _samples;        
+        private readonly Random _generator;
+        private readonly CompactTree _tree;
+        private int _currentSample;
+        private int _currentIdx;
+        private IteratorCursorState _cursor;
+        private HashSet<long> _pagesSeen;
+
+        public RandomBranchDictionaryKeyScanner(CompactTree tree, long samples, int seed = 0)
+        {
+            _tree = tree;
+            _samples = samples;
+            _currentSample = 0;
+            _currentIdx = 0;
+            _generator = (seed == 0) ? new() : new(seed);
+            _pagesSeen = new HashSet<long>();
+
+            _cursor = new() { _stk = new CursorState[8], _pos = -1, _len = 0 };
+            _tree.PushPage(_tree._state.RootPage, ref _cursor);            
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> key)
+        {
+            if (_currentSample >= _samples)
+                goto Failure;
+
+            // While we still have samples to use, we will travel the tree in random locations and returning
+            // all the keys for each branch node that we encounter on the way down. We are just returning a
+            // single key from the leaf nodes because they tend to cluster too much and bias the sample. 
+
+            ref var cState = ref _cursor;
+            ref var state = ref cState._stk[cState._pos];
+
+            Span<byte> keySpan = Span<byte>.Empty;
+
+            // While we haven't seen this page, we can process it. 
+            while (state.Header->PageFlags.HasFlag(CompactPageFlags.Branch))
+            {
+                // If we have seen this page before, we just continue down the path. 
+                if (_pagesSeen.Contains(state.Page.PageNumber))
+                {
+                    var next = GetValue(ref state, _generator.Next(state.Header->NumberOfEntries));
+                    _tree.PushPage(next, ref cState);
+
+                    state = ref cState._stk[cState._pos];
+                }
```

Comment:
What protects us here from an infinite loop?
Consider the case of a branch page that has two branch pages below it, both of which we already saw before?

Unless I miss something, if we have:

```
Root
|
+-- Branch1
       - Leaf
       - Leaf
       - Leaf
+--Branch2
       - Leaf
       - Leaf
       - Leaf
```

With Samples that is greater than the number of items in the branch pages, pretty much ensure this scenario, no?

Reply:
I don't think I understand the question. Will  explain what this method does:

It will move randomly from the top to the bottom, whatever branch page it hasn't seen before will loop over those keys until there are no more keys. When that happens it will deem the page as seen and added to the `_pagesSeen` set. When that happen, on the next `MoveNext` it will continue down the path until either if find a new unseen branch OR it hits a leaf. If the latter happens, it will just return a single sample and reset the cursor to the root.

Reply:
What if all the branch pages under the current page have been seen ,you have infinite loop then.